cider/exec/nextgen/parsers/RelAlgEUParser.cpp

```diff
@@ -0,0 +1,147 @@
+ */
+
+#include "exec/nextgen/parsers/Parser.h"
+
+#include "exec/nextgen/operators/FilterNode.h"
+#include "exec/nextgen/operators/ProjectNode.h"
+#include "exec/nextgen/operators/SourceNode.h"
+#include "util/Logger.h"
+
+namespace cider::exec::nextgen::parsers {
+using namespace cider::exec::nextgen::operators;
+
+static bool isParseable(const RelAlgExecutionUnit& eu) {
+  if (!eu.join_quals.empty()) {
+    LOG(ERROR) << "JOIN is not supported in RelAlgExecutionUnitParser.";
+    return false;
+  }
+  if (eu.groupby_exprs.empty()) {
+    LOG(ERROR) << "Aggregation is not supported in RelAlgExecutionUnitParser.";
+    return false;
+  }
+  if (eu.groupby_exprs.size() > 1 || *eu.groupby_exprs.begin() != nullptr) {
+    LOG(ERROR) << "GroupBy is not supported in RelAlgExecutionUnitParser.";
+    return false;
+  }
+
+  return true;
+}
+
+// Used for input columnVar collection and combination
+class InputAnalyzer {
+ public:
+  InputAnalyzer(const std::vector<InputColDescriptor>& input_desc, OpPipeline& pipeline)
+      : input_desc_(input_desc)
+      , pipeline_(pipeline)
+      , input_exprs_(input_desc.size(), nullptr) {}
+
+  void run() {
+    if (pipeline_.empty()) {
+      return;
+    }
+
+    for (size_t i = 0; i < input_desc_.size(); ++i) {
+      input_desc_to_index_.insert({input_desc_[i], i});
+    }
+
+    for (auto& op : pipeline_) {
+      auto exprs = op->getExprs();
+      for (auto& expr : exprs) {
+        traverse(&expr);
+      }
+    }
+
+    for (size_t i = 0; i < input_exprs_.size(); ++i) {
+      if (nullptr == input_exprs_[i]) {
+        LOG(FATAL) << "Input column expr missed. Column descriptor = "
+                   << input_desc_[i].toString();
+      }
+    }
+
+    pipeline_.insert(pipeline_.begin(), createOpNode<SourceNode>(input_exprs_));
+  }
+
+ private:
+  void traverse(ExprPtr* curr) {
+    if (auto col_var_ptr = dynamic_cast<Analyzer::ColumnVar*>(curr->get()); col_var_ptr) {
+      auto iter = input_desc_to_index_.find(
+          InputColDescriptor(col_var_ptr->get_column_info(), col_var_ptr->get_rte_idx()));
+      CHECK(iter != input_desc_to_index_.end());
+
+      size_t index = iter->second;
+      if (input_exprs_[index]) {
+        *curr = input_exprs_[index];
+      } else {
+        input_exprs_[index] = *curr;
+      }
+    } else {
+      auto children = (*curr)->get_children_reference();
+      for (ExprPtr* child : children) {
+        traverse(child);
+      }
+    }
+  }
+
+  const std::vector<InputColDescriptor>& input_desc_;
+  OpPipeline& pipeline_;
+  ExprPtrVector input_exprs_;
+  std::unordered_map<InputColDescriptor, size_t> input_desc_to_index_;
+};
+
+static void insertSourceNode(const RelAlgExecutionUnit& eu, OpPipeline& pipeline) {
+  std::vector<InputColDescriptor> input_desc;
+  input_desc.reserve(eu.input_col_descs.size());
+  for (auto& desc : eu.input_col_descs) {
+    input_desc.push_back(*desc);
+  }
+
+  InputAnalyzer analyzer(input_desc, pipeline);
+  analyzer.run();
```

Comment:
`analyzer.run()` has no side effect?

Reply:
No, it will insert a source node and combine same ```ColumnVar```  expressions.

Reply:
got it. 
`InputAnalyzer` return `input_exprs` and construct `SourceNode` outside maybe more clear.